File: Python/_0000_study/_leetcode/_swordoffer/_053_search/Solution.py
# coding=utf-8
# Date: 2021/7/16 9:48
from typing import List


# 执行用时：32 ms, 在所有 Python3 提交中击败了95.55%的用户
# 内存消耗：15.6 MB, 在所有 Python3 提交中击败了26.71%的用户
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 0:
            return 0
        l, r = 0, len(nums) - 1
        while l < r:
            mid = l + r >> 1
            if nums[mid] >= target:
                r = mid
            else:
                l = mid + 1
        if nums[r] != target:
            return 0

        L = l
        l, r = 0, len(nums) - 1
        while l < r:
            mid = l + r + 1 >> 1
            if nums[mid] <= target:
                l = mid
            else:
                r = mid - 1
        return r - L + 1


if __name__ == "__main__":
    print(Solution().search([5, 7, 7, 8, 8, 10], 8))  # 2

File: Java/src/_0000_study/_leetcode/_swordoffer/_053_search/Solution.java
package _0000_study._leetcode._swordoffer._053_search;

/**
 * Date: 2021/7/16 9:45
 * Content:
 * 执行用时：0 ms, 在所有 Java 提交中击败了100.00%的用户
 * 内存消耗：41.2 MB, 在所有 Java 提交中击败了72.79%的用户
 */
public class Solution {

    public int search(int[] nums, int target) {

        if (nums.length == 0) return 0;

        int n = nums.length;
        int l = 0, r = n - 1;
        while (l < r) {
            int mid = l + r >> 1;
            if (nums[mid] >= target) r = mid;
            else l = mid + 1;
        }

        if (nums[l] != target) return 0;
        int L = l;
        l = 0;
        r = n - 1;
        while (l < r) {
            int mid = l + r + 1 >> 1;
            if (nums[mid] <= target) l = mid;
            else r = mid - 1;
        }
        return r - L + 1;
    }

    public static void main(String[] args) {

        int[] nums = {5, 7, 7, 8, 8, 10};
        System.out.println((new Solution()).search(nums, 8));  // 2
    }
}
